<!DOCTYPE html>
<html>
<head>
  <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
  <title>The source code</title>
  <link href="../resources/prettify/prettify.css" type="text/css" rel="stylesheet" />
  <script type="text/javascript" src="../resources/prettify/prettify.js"></script>
  <style type="text/css">
    .highlight { display: block; background-color: #ddd; }
  </style>
  <script type="text/javascript">
    function highlight() {
      document.getElementById(location.hash.replace(/#/, "")).className = "highlight";
    }
  </script>
</head>
<body onload="prettyPrint(); highlight();">
  <pre class="prettyprint lang-js">describe(&quot;Ext.Error&quot;, function() { 
    var global;

    beforeEach(function() {
        global = Ext.global;

        // mock the console to avoid logging to the real console during the tests
        Ext.global = {
            console: {
                dir: function(s) {
                    return s;
                },
                log: function(s) {
                    return s;
                },                
                error: function(s) {
                    return s;
                },
                warn: function(s) {
                    return s;
                }
            }
        };
    });

    afterEach(function() {
        Ext.global = global;
    });

    describe(&quot;raising an error via Ext.Error.raise&quot;, function() {

        describe(&quot;passing a string&quot;, function() {
    
            it(&quot;should throw an error with a msg property&quot;, function() {
                var error;
                try {
                    Ext.Error.raise('foo');
                }
                catch (err) {
                    error = err;
                }
                expect(error.msg).toEqual('foo');
            });
        
            it(&quot;should log an error to the console&quot;, function() {
                spyOn(Ext.global.console, 'error');
                try {
                    Ext.Error.raise('foo');
                } 
                catch (err) {}
                expect(Ext.global.console.error).toHaveBeenCalledWith('[E] foo');
            });
        
            it(&quot;should log the error object to the console&quot;, function() {
                spyOn(Ext.global.console, 'dir').andCallFake(function(err){
                    expect(err.msg).toEqual('foo');
                });
                try {
                    Ext.Error.raise('foo');
                } 
                catch (err) {}
            });
        
            it(&quot;should do nothing when Ext.Error.ignore = true&quot;, function() {
                spyOn(Ext.global.console, 'warn');
            
                Ext.Error.ignore = true;
                try {
                    Ext.Error.raise('foo');
                } 
                catch (err) {
                    expect('Error should not have been caught').toBe(true);
                }
                expect(Ext.global.console.warn).not.toHaveBeenCalled();
                Ext.Error.ignore = false;
            });
        
            it(&quot;should not throw an error if handled by Ext.Error.handle&quot;, function() {
                spyOn(Ext.global.console, 'warn');
            
                var origHandle = Ext.Error.handle;
                Ext.Error.handle = function(err) {
                    expect(err.msg).toEqual('foo');
                    return true;
                }
                try {
                    Ext.Error.raise('foo');
                } 
                catch (err) {
                    expect('Error should not have been caught').toBe(true);
                }
                expect(Ext.global.console.warn).not.toHaveBeenCalled();
                Ext.Error.handle = origHandle;
            });
        });
    
        describe(&quot;passing an object with a msg property&quot;, function() {
    
            it(&quot;should throw an error with a msg property&quot;, function() {
                var error;
                try {
                    Ext.Error.raise({msg: 'foo'});
                }
                catch (err) {
                    error = err;
                }
                expect(error.msg).toEqual('foo');
            });
        
            it(&quot;should log an error to the console&quot;, function() {
                spyOn(Ext.global.console, 'error');
                try {
                    Ext.Error.raise({msg: 'foo'});
                } 
                catch (err) {}
                expect(Ext.global.console.error).toHaveBeenCalledWith('[E] foo');
            });
        
            it(&quot;should log the error object to the console&quot;, function() {
                spyOn(Ext.global.console, 'dir').andCallFake(function(err){
                    expect(err.msg).toEqual('foo');
                });
                try {
                    Ext.Error.raise({msg: 'foo'});
                } 
                catch (err) {}
            });
                            
            it(&quot;should do nothing when Ext.Error.ignore = true&quot;, function() {
                spyOn(Ext.global.console, 'warn');
            
                Ext.Error.ignore = true;
                try {
                    Ext.Error.raise({msg: 'foo'});
                } 
                catch (err) {
                    expect('Error should not have been caught').toBe(true);
                }
                expect(Ext.global.console.warn).not.toHaveBeenCalled();
                Ext.Error.ignore = false;
            });
        
            it(&quot;should not throw an error if handled by Ext.Error.handle&quot;, function() {
                spyOn(Ext.global.console, 'warn');
            
                var origHandle = Ext.Error.handle;
                Ext.Error.handle = function(err) {
                    expect(err.msg).toEqual('foo');
                    return true;
                }
                try {
                    Ext.Error.raise({msg: 'foo'});
                } 
                catch (err) {
                    expect('Error should not have been caught').toBe(true);
                }
                expect(Ext.global.console.warn).not.toHaveBeenCalled();
                Ext.Error.handle = origHandle;
            });
        });
    
        describe(&quot;passing an object with custom metadata&quot;, function() {
    
            it(&quot;should throw an error with matching metadata&quot;, function() {
                var error;
                try {
                    Ext.Error.raise({
                        msg: 'Custom error',
                        data: {
                            foo: 'bar'
                        }
                    });
                }
                catch (err) {
                    error = err;
                }
                expect(error.msg).toEqual('Custom error');
                expect(error.data).not.toBe(null);
                expect(error.data.foo).toEqual('bar');
            });
        
            it(&quot;should log the complete metadata to the console&quot;, function() {
                spyOn(Ext.global.console, 'dir').andCallFake(function(err){
                    expect(err.msg).toEqual('Custom error');
                    expect(err.data).not.toBe(null);
                    expect(err.data.foo).toEqual('bar');
                });
                try {
                    Ext.Error.raise({
                        msg: 'Custom error',
                        data: {
                            foo: 'bar'
                        }
                    });
                } 
                catch (err) {}
            });
        });
    
        describe(&quot;originating from within a class defined by Ext&quot;, function() {
    
            Ext.define('CustomClass', {
                doSomething: function(o){
                    Ext.Error.raise({
                        msg: 'Custom error',
                        data: o,
                        foo: 'bar'
                    });
                }
            });
            var customObj = Ext.create('CustomClass');
        
            it(&quot;should throw an error containing the source class and method&quot;, function() {
                var error;
                try {
                    customObj.doSomething({
                        extraData: 'extra'
                    });
                }
                catch (err) {
                    error = err;
                }
                expect(error.msg).toEqual('Custom error');
                expect(error.sourceClass).toEqual('CustomClass');
                expect(error.sourceMethod).toEqual('doSomething');
                expect(error.toString()).toBe('CustomClass.doSomething(): Custom error');
            });
        
            it(&quot;should log the complete metadata to the console&quot;, function() {
                spyOn(Ext.global.console, 'dir').andCallFake(function(err){
                    expect(err.msg).toEqual('Custom error');
                    expect(err.sourceClass).toEqual('CustomClass');
                    expect(err.sourceMethod).toEqual('doSomething');
                    expect(err.data).not.toBe(null);
                    expect(err.data.extraData).not.toBe(null);
                    expect(err.data.extraData).toEqual('extra');
                    expect(err.foo).toEqual('bar');
                });
                try {
                    customObj.doSomething({
                        extraData: 'extra'
                    });
                } 
                catch (err) {
                }
            });
        });
    });
    
    describe(&quot;Throwing an an Ext.Error directly intantiated&quot;, function() {
        describe(&quot;Passing an string as constructor argument&quot;, function() {
           it(&quot;should contain a msg property with the given string as value&quot;, function() {
              expect(function() {
                  throw new Ext.Error(&quot;expected message&quot;);
              }).toRaiseExtError(&quot;expected message&quot;);
           });
        });
     });
    
    xdescribe(&quot;Ext.deprecated&quot;, function() {
       // failing only on CI
       it(&quot;should return a function that raises an error with the given suggestion&quot;, function() {
          Ext.ClassManager.enableNamespaceParseCache = false;
          Ext.define(&quot;Test.ThisClassContainsADeprecatedMethod&quot;, {
             deprecatedMethod : Ext.deprecated(&quot;use another function&quot;)
          });
          expect(function() {
              new Test.ThisClassContainsADeprecatedMethod().deprecatedMethod();
          }).toThrow('The method &quot;Test.ThisClassContainsADeprecatedMethod.deprecatedMethod&quot; has been removed. use another function');
          
          try {
              delete Test;
          } catch(e) { }
          Ext.ClassManager.enableNamespaceParseCache = true;
       });
    });
});
</pre>
</body>
</html>
